### PR TITLE
Fix `RestartWhenAppExited` flag being set before confirmation step

### PR DESCRIPTION
**Pull request by peppy**
Addresses https://github.com/ppy/osu/discussions/21003.

### CODE REVIEW

**Review by bdach — DISMISSED** [review]
I have no earthly idea how I'm supposed to be testing this, but at least the code looks sound.

File: osu.Game/OsuGame.cs [hunk 1]
```diff
@@ -651,10 +651,14 @@ public void PerformFromScreen(Action<IScreen> action, IEnumerable<Type> validScr
...
 
-        public override void AttemptExit()
+        public override void AttemptExit(Action invokeWhenExiting = null)
         {
             // Using PerformFromScreen gives the user a chance to interrupt the exit process if needed.
-            PerformFromScreen(menu => menu.Exit());
+            PerformFromScreen(menu =>
+            {
+                invokeWhenExiting?.Invoke();
+                menu.Exit();
```
Comment (Joehuu) [inline osu.Game/OsuGame.cs:660]: Tested with `osu-deploy` by hardcoding version to an older date and making a custom build that updates to the latest release.

This doesn't seem to work. `menu.Exit()` calls `MainMenu.OnExiting()` / opens the confirmation dialog:
https://github.com/ppy/osu/blob/9724c6bb5350bb4e19731d4baadf6c96ca9c1514/osu.Game/Screens/Menu/MainMenu.cs#L280-L299

which means it'll still set `RestartWhenAppExited` before the dialog is even opened.

Comment (peppy) [reply osu.Game/OsuGame.cs:660]: Right.. that's one level higher than I expected. Bit of a pain to get this right. Maybe the menu should just set the flag itself somehow..

**Comment by peppy** [comment]
No idea how to fix this in a sane way. Probably fine to just leave it broken, very edge case.

**Comment by LittleEndu** [comment]
Looking at squirrel source, isn't it possible to kill the Update.exe that's waiting for restart?

Yeah this is an edge case, but very weird UX. If I just misclick on the update ready notification then it's weird to see the app start again.

**Comment by peppy** [comment]
If it's implemented, it should be done without killing a process. You're welcome to give it a try, but it's going to be a difficulty one to structure in a nice way.

**Pull request closed without merging**